### ISSUE TITLE
Meson with Visual Studio

### ISSUE THREAD

**Issue by Davido71** [haasn/libplacebo#100]
Trying to build libplacebo with Visual Studio 2019 but Meson is not detecting Vulkan. I do have Vulkan SDK installed and is detected by projects that use cmake. Mako is also installed ('pip3 install mako'). According to the Meson documentation (https://mesonbuild.com/Dependencies.html) VULKAN_SDK environment variable is all I need.

Here is my bat file.

call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Auxiliary\Build\vcvarsall.bat" amd64
set VULKAN_SDK=C:\VulkanSDK\1.2.170.0     REM Redundant, also in my environment variables  
meson ./build -Dopengl=disabled

Meson output:
![image](https://user-images.githubusercontent.com/61202402/116907990-9a612500-abff-11eb-8c82-d1d72e17ab78.png)

Any advice or direction would be much appreciated.
Thanks!

**Comment by 1480c1** [comment]
From what I know, meson looks for `dependency()`s through either pkg-config, cmake, or from meson itself, and I'm not sure how cmake projects normally do it, but I am assuming vulkan's sdk doesn't provide a normal FindVulkan.cmake file where you can just use `find_package(VULKAN vulkan)` nor a pkg-config file, so I'm not sure what would be the best way to go about trying to find vulkan on windows with msvc without pkgconf etc

**Comment by haasn** [comment]
> unknown version, but need: ['>=1.0.42']

This makes it seem like a meson bug, though a full log would help to know for sure. That said, given that 1.0.42 isn't exactly a high dependency, we can most likely work around it by simply removing the version requirement.

Does this patch fix it?

```diff
diff --git a/src/meson.build b/src/meson.build
index 0d8775d..4a368ab 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -5,7 +5,7 @@ libm = cc.find_library('m', required: false)
 build_deps = [ libm ]
 test_deps = []
 
-vulkan = dependency('vulkan', version: '>=1.0.42', required: get_option('vulkan'))
+vulkan = dependency('vulkan', required: get_option('vulkan'))
 opengl = dependency('epoxy', version: '>=1.4.0', required: get_option('opengl'))
 shaderc = dependency('shaderc', version: '>=2018.0', required: get_option('shaderc'))
 
```

**Comment by Davido71** [comment]
@haasn  Thanks the quick response and patch.  I was able to get past the Vulkan version issue. The next dependency it complained about was shaderc. I got past this by using direct paths to library and include.

shaderc = declare_dependency(link_with : static_library('C:\\VulkanSDK\\1.2.170.0\\Lib\\shaderc.lib'), include_directories : 'C:\\VulkanSDK\\1.2.170.0\\Include\\shaderc')

At this point meson seems to complete but when I run ninja to build I get the following error:

Found ninja-1.10.2 at "C:\Program Files\Meson\ninja.EXE"
Traceback (most recent call last):
  File "mesonbuild\mesonmain.py", line 132, in run
  File "mesonbuild\msetup.py", line 279, in run
  File "mesonbuild\msetup.py", line 183, in generate
  File "mesonbuild\msetup.py", line 245, in _generate
  File "mesonbuild\backend\ninjabackend.py", line 533, in generate
  File "mesonbuild\backend\ninjabackend.py", line 691, in generate_target
  File "os.py", line 215, in makedirs
  File "os.py", line 215, in makedirs
  File "os.py", line 215, in makedirs
  [Previous line repeated 1 more time]
  File "os.py", line 225, in makedirs
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Projects\\_3rdPartyBuilds\\libplacebo\\build\\src\\libC:'
ninja: error: loading 'build.ninja': The system cannot find the file specified.

I'm thinking that meson is more Linux centric. Thinking about converting the project to cmake which I'm familiar with.
Thanks!

**Comment by haasn** [comment]
> I'm thinking that meson is more Linux centric. Thinking about converting the project to cmake which I'm familiar with.

Probably easier would be to use [WSL](https://docs.microsoft.com/en-us/windows/wsl/about) to run meson in a Linux-like environment.

That said, I'm surprised that this doesn't work out of the box. Meson is *supposed* to support Windows, so this is definitely another meson bug. I guess meson just lacks windows users/developers interested in filing issues and submitting patches.

**Comment by 1480c1** [comment]
tbh, most of the meson users on windows more often use msys2+mingw-w64 gcc and I think msvc with it is very rare

**Comment by Davido71** [comment]
> I think msvc with it is very rare
That's interesting, I may have to rethink my life. :) 

Thanks for the tips! I do use msys2 to build ffmpeg and other projects. I'll attempt to build using msys2. I know that msys2 has a libplacebo package (mingw-w64-x86_64-libplacebo) but I need to generate a .def file so I can use it to create a .lib file (visual studio again).

**Comment by haasn** [comment]
> but I need to generate a .def file so I can use it to create a .lib file (visual studio again).

MinGW apparently comes with a tool to generate a .def from a .dll (gendef), have you tried that?

**Comment by Davido71** [comment]
> MinGW apparently comes with a tool to generate a .def from a .dll (gendef), have you tried that?

I tried gendef and it indeed build a .def file which in turn I was able to build a .lib file. 

Here are the steps if anyone wants to try it.
From the msys2 x64 prompt:
gendef /mingw64/bin/libplacebo-104.dll

From the Visual Studio x64 Native Tools Command prompt:
lib /def:libplacebo-104.def /out:libplacebo-104.lib /machine:x64

Thanks for you help!
